File: src/main/java/axirassa/webapp/services/AxirassaSecurityService.java

package axirassa.webapp.services;

import org.apache.shiro.subject.Subject;
import org.hibernate.Session;

import axirassa.model.EntityWithUser;
import axirassa.model.UserEntity;
import axirassa.webapp.services.exceptions.AxirassaSecurityException;

public interface AxirassaSecurityService {

	public abstract String getEmail();


	/**
	 * Returns the {@link UserEntity} associated with the authenticated user, or
	 * null if the user is not authenticated.
	 */
	public abstract UserEntity getUserEntity(Session database);


	public abstract boolean isUser();


	public abstract boolean isGuest();


	public abstract boolean isAuthenticated();


	public abstract Subject getSubject();


	/**
	 * Throws an {@link AxirassaSecurityException} if the given entity returns
	 * null from {@link EntityWithUser#getUser()} or if the user returned does
	 * not match the user currently logged in.
	 * 
	 * If the given entity is null, no exception is thrown. (good luck getting
	 * any data out of it)
	 * 
	 * @throws AxirassaSecurityException
	 */
	public abstract void verifyOwnership(EntityWithUser entity) throws AxirassaSecurityException;

}

File: src/main/java/axirassa/webapp/services/AxirassaSecurityService.scala
package axirassa.webapp.services

import axirassa.dao.UserEmailAddressDAO
import axirassa.dao.UserDAO
import org.hibernate.Session
import org.tynamo.security.services.SecurityService
import org.apache.tapestry5.ioc.annotations.Inject
import axirassa.model.EntityWithUser
import org.apache.shiro.subject.Subject
import axirassa.model.UserEntity


class AxirassaSecurityException extends Exception("Unauthorized access attempt")


trait AxirassaSecurityService {
    def getEmail : String
    def getUserEntity : UserEntity
    def isUser : Boolean
    def isGuest : Boolean
    def isAuthenticated : Boolean
    def getSubject : Subject
    def verifyOwnership(entity : EntityWithUser)
}

class AxirassaSecurityServiceImpl extends AxirassaSecurityService {
    @Inject
    var security : SecurityService = _

    @Inject
    var database : Session = _

    @Inject
    var userDAO : UserDAO = _

    @Inject
    var userEmailAddressDAO : UserEmailAddressDAO = _

    override def isAuthenticated = security.isAuthenticated
    override def isGuest = security.isGuest
    override def isUser = security.isUser
    override def getEmail = security.getSubject.getPrincipal.asInstanceOf[String]
    override def getSubject = security.getSubject

    override def verifyOwnership(entity : EntityWithUser) {
        if (entity.getUser == null)
            throw new AxirassaSecurityException

        val user = entity.getUser
        val email = userEmailAddressDAO.getPrimaryEmail(user)

        // user has no primary e-mail (weird)
        if (email.isEmpty)
            throw new AxirassaSecurityException

        if (getEmail != email.get)
            throw new AxirassaSecurityException
    }

    override def getUserEntity = {
        if (getEmail == null)
            null
        else {
            val user = userDAO.getUserByEmail(getEmail)
            if (user.isEmpty)
                throw new AxirassaSecurityException

            user.get
        }
    }
}